Sming/SmingCore/Network/HttpRequest.cpp

```diff
@@ -165,61 +183,35 @@ HttpParseResult HttpRequest::parsePostData(HttpServer *server, pbuf* buf)
 		}
 		requestPostParameters = new HashMap<String, String>();
 		start = headerEnd + 4;
-		combinePostFrag = false;
-	}
-	else if (combinePostFrag)
-	{
-		String cur = requestPostParameters->keyAt(requestPostParameters->count() - 1);
-		debugf("Continue POST frag %s", cur.c_str());
-		int delimItem = NetUtils::pbufFindChar(buf, '&', 0);
-		if (delimItem == -1)
-			delimItem = buf->tot_len;
-		else
-			combinePostFrag = false;
-		String itemValue = NetUtils::pbufStrCopy(buf, 0, delimItem);
-		//debugf("Continue POST len %d", itemValue.length());
-		char* buf = uri_unescape(NULL, 0, itemValue.c_str(), -1);
-		itemValue = buf;
-		free(buf);
-		(*requestPostParameters)[cur] += itemValue;
-		start = delimItem + 1;
-		postDataProcessed += start;
+		tmpbuf = tmpbuf.substring(start, tmpbuf.length());
 	}
 
-	bool notFinished = extractParsingItemsList(buf, start, buf->tot_len, '&', ' ', requestPostParameters);
-	if (notFinished)
-		combinePostFrag = true; // continue reading this parameter value
-	//TODO: continue for param name
-	postDataProcessed += buf->tot_len - start;
+	tmpbuf = extractParsingItemsList(tmpbuf, start, tmpbuf.length(), '&', ' ', requestPostParameters);
+	postDataProcessed += buf->tot_len - start ;
 
 	if (postDataProcessed == getContentLength())
 		return eHPR_Successful;
 	else
+
 		return eHPR_Wait;
 }
 
-bool HttpRequest::extractParsingItemsList(pbuf* buf, int startPos, int endPos, char delimChar, char endChar,
+String HttpRequest::extractParsingItemsList(String& buf, int startPos, int endPos, char delimChar, char endChar,
 													HashMap<String, String>* resultItems)
 {
-	bool continued = false;
+
 	int delimItem, nextItem, startItem = startPos;
-	while (startItem < endPos)
+	do
 	{
-		delimItem = NetUtils::pbufFindStr(buf, "=", startItem);
-		if (delimItem == -1 || delimItem > endPos) break;
-		nextItem = NetUtils::pbufFindChar(buf, delimChar, delimItem + 1);
-		if (nextItem == -1)
-			nextItem = NetUtils::pbufFindChar(buf, endChar, delimItem + 1);
-		if (nextItem > endPos) break;
-
-		if (nextItem == -1)
-		{
-			nextItem = endPos;
-			continued = true;
-		}
+		//debugf("item %i  - delim %i - next %i", startItem, delimItem, nextItem);
+		delimItem = buf.indexOf("=", startItem);
+		nextItem = buf.indexOf(delimChar, startItem);
+		if (nextItem == -1) nextItem = buf.indexOf(endChar, delimItem+1);
+		if (nextItem == -1) nextItem = endPos;
+		if (nextItem > endPos || delimItem == -1) nextItem = endPos; break;
 
-		String ItemName = NetUtils::pbufStrCopy(buf, startItem, delimItem - startItem);
-		String ItemValue = NetUtils::pbufStrCopy(buf, delimItem + 1, nextItem - delimItem - 1);
+		String ItemName = buf.substring(startItem, delimItem);
+		String ItemValue = buf.substring(delimItem+1, nextItem);
```

Comment:
Short answer since I am on mobile. Substring uses absolute numbers
substring(Startposition, endposition)  while pbuf used (Startposition,
length). So this should be correct
On Mar 1, 2016 16:10, "Jose Barros" <notifications@github.com> wrote:

> In Sming/SmingCore/Network/HttpRequest.cpp
> <https://github.com/SmingHub/Sming/pull/622#discussion_r54579516>:
>
> >
> > -		String ItemName = NetUtils::pbufStrCopy(buf, startItem, delimItem - startItem);
> > -		String ItemValue = NetUtils::pbufStrCopy(buf, delimItem + 1, nextItem - delimItem - 1);
> > +		String ItemName = buf.substring(startItem, delimItem);
> > +		String ItemValue = buf.substring(delimItem+1, nextItem);
>
> Shouldn't it be buf.substring(delimItem + 1, nextItem - delimItem - 1) ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SmingHub/Sming/pull/622/files#r54579516>.
>